_posts/2022-07-07-premature-dispsosal.md

```diff
@@ -0,0 +1,87 @@
+
+## Unpredictable behavior
+
+Whether `DoSomething` runs into this problem depends on what it does internally. As with throwing Exceptions, the behavior between methods that return Tasks but are actually synchronous (e.g. `Task SyncMethod() => Task.CompletedTask;`) and truly asynchronous methods (e.g. `async Task AsyncMethod() => await Task.Yield();`) varies. Consider these two implementations of `DoSomething`:
+
+```csharp
+static Task DoSomething(MyDisposableClass myDisposableClass)
+{
+    myDisposableClass.DoStuff();
+    return Task.CompletedTask;
+}
+```
+
+and
+
+```csharp
+static async Task DoSomething(MyDisposableClass myDisposableClass)
+{
+    await Task.Yield();
+
+    myDisposableClass.DoStuff();
+}
+```
+
+The first implementation will run without any exception in the intially shown `Main` method because `DoSomething` will run to completion before it returns to `Main` to dispose the passed object reference. The second version will throw an `ObjectDisposedException` because `Task.Yield()` will trigger the return to the caller and finish the remaining code of the asynchronous method later. We need to assume that any real `Task` returning method will have actual asynchronous code exception (like the 2nd version) though. In reality, a single method can even show both behaviors, completely depending on runtime conditions:
+
+```csharp
+static Task DoSomething(MyDisposableClass myDisposableClass)
+{
+    if (myDisposableClass.Validate)
+    {
+        return InvokeWithExpensiveValidation();
+
+        async Task InvokeWithExpensiveValidation()
+        {
+            await ExpensiveValidationAsync();
+            myDisposableClass.DoStuff();
+        }
+    }
+    else
+    {
+        myDisposableClass.DoStuff();
+        return Task.CompletedTask;
+    }
+}
+```
+
+An implementation like this can avoid the cost of asynchronous methods when it is expected that the async paths is only rarely used. This makes it even harder to notice the initial mistake to not `await` the method call because the code might work just fine most of time time, until it doesn't.
+
+## Conclusion
+
+The `using var` language feature is fantastic and much more convenient than having to define `using(...){ ... }` blocks all over the place. But in combination with asynchronous code, it can be fairly easy to miss the additional `using` keyword (imagine a much larger method where the disposable instance is created at the very beginning). There is currently no compiler or code analyzer that will detect such cases of "premature disposals" (or rather missing `await`s), so keep an eye out for this and make sure to fully `await` any method that receives a reference to disposable objects.
```

Comment:
```suggestion
The `using var` language feature is fantastic and much more convenient than having to define `using(...){ ... }` blocks all over the place. But with asynchronous code, it can be easy to miss the additional `using` keyword (imagine a scenario where the disposable instance is created at the very beginning).

No compiler or code analyzer detects such cases of "premature disposals" (or missing `await`s). It's crucial to keep an eye out for this and properly `await` any method that receives a reference to disposable objects.
```